### DISCUSSION TITLE
How to disambiguate invalid parens, f.e. `();`, from valid arrow function, f.e. `() => {}`

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#1056] [Q&A] [answered]
We have the following expression to detect "invalid" empty parens by having the validator mark the valid syntax as invalid:

https://github.com/bytescript/bytescript/blob/bddbe1b452cbcba256d761a3f391f9ea63c8ebb9/src/language-server/bytescript.langium#L92-L93
> 
> ```js
> InvalidParenthesis:
> 	empty?="(" ")";
> ```

 I'm also trying to add arrow function syntax into the grammar,

https://github.com/bytescript/bytescript/blob/bddbe1b452cbcba256d761a3f391f9ea63c8ebb9/src/language-server/bytescript.langium#L119
```js
ArrowFunctionExpression:
	Parameters (":" ReturnType)? "=>" body=(ArrowReturnExpression | Block);
```

where `Parameters` is a fragment with `"(" ... ")"`, however it conflicts with the `InvalidParenthesis` rule, so when I try to write an arrow function in my sample code, I get an error regarding the invalid parens instead.

For example, this is invalid code that the validator should flag,

```js
();
```

while this is valid code:

```js
const f = () => {}
```

How can we achieve having both of those things in the language?

**Comment by msujew** [comment, marked as answer]
It seems like `ArrowFunctionExpression` is currently unused, right? I've tried your current `main` with a minor change:

```diff
FunctionDeclaration:
-	OriginalFunctionDeclaration | GeneratorFunctionDeclaration;
+	OriginalFunctionDeclaration | GeneratorFunctionDeclaration | ArrowFunctionExpression;
```

That seems to work as expected:

![image](https://github.com/langium/langium/assets/4377073/3d9e74de-a0bb-4e8f-82c7-b759bd3b6318)

**Reply by trusktr** [reply]
Ah, ok! Simple things make a difference. Thanks!
